### ISSUE TITLE
Build installer packages automatically

### ISSUE THREAD

**Issue by bittner** [cool-RR/PythonTurtle#107]
Following up on issues https://github.com/cool-RR/PythonTurtle/pull/105#issuecomment-421965482, #106 and #90, I'd like to start collecting information on how to (best) do packaging for all platforms PythonTurtle runs on. When we know how to perform packaging manually it will be an easy task to automate it all.

In other words, we need
1. a list of **target package types** (e.g. MSI for Windows, macOS AppImage, Debian DEB, RPM, Snap), and
1. a **complete list of concrete, detailed steps** (*read:* shell commands) of how each package type is built, from source code living in source control only.

NOTE: *If writing this up in this issue proves uncomfortable we may want to use the [project Wiki](https://github.com/cool-RR/PythonTurtle/wiki) to help us move along.*

**Comment by bittner** [comment]
We [currently seem](https://github.com/cool-RR/PythonTurtle/tree/master/packaging_scripts/plumbing) to use [py2exe](http://py2exe.org/) and [py2app](https://py2app.readthedocs.io/) for Windows and macOS.

According to a [recent talk at PyCon Otto](https://www.pycon.it/media/conference/slides/pachettizzare-applicazioni-python-in-un-singolo-file-binario.pdf) we should switch to [PyInstaller](http://www.pyinstaller.org/), which allows building stand-alone executables for Windows, macOS and GNU/Linux alike.

**Comment by cool-RR** [comment]
I agree we should definitely use PyInstaller instead of py2exe and py2app.

And I agree that we need a reliable and automatic build process.

**Comment by bittner** [comment]
We should also not forget about building the [PythonTurtle](https://pypi.org/project/PythonTurtle/) Python package and push it to PyPI when releasing.

I'd also suggest to trigger releases by pushing a Git tag (with the version number) to the GitHub repository. The installers may then be downloadable via [the release page](https://github.com/cool-RR/PythonTurtle/releases). I hope that's all possible in a seamless way.

**Comment by cool-RR** [comment]
I agree, though that kind of automation is a bit of a high goal, if we even get to the point where we could do a release, I'd be ecstatic.

**Comment by bittner** [comment]
>  if we even get to the point where we could do a release, I'd be ecstatic.

Then prepare your space suit!  :smile:

**Comment by bittner** [comment]
As suspected PyInstaller will require to [run on the infrastructure it should build the installer for](https://pyinstaller.readthedocs.io/en/stable/usage.html#supporting-multiple-operating-systems) (e.g. on Windows for a Windows installer, on a Mac for a macOS installer).

I have neither a Windows and Mac. Hence, I need to do the cumbersome finding-out how building is done on AppVeyor. Also, we need a macOS with Python 3 installed (and cross your fingers that [the documentation is only outdated](https://pyinstaller.readthedocs.io/en/stable/installation.html#installing-in-mac-os-x) and we can also build with Python 3).

I'll probably best start with automating building a Python package (and upload it to PyPI automatically). IIUC, a `setup.py` is needed for building with (some features of) PyInstaller, anyway.

**Comment by cool-RR** [comment]
I added you as a maintainer on PyPI.

Regarding PyInstaller, I wonder whether a VM with ReactOS would work for generating a Windows installer.

**Comment by bittner** [comment]
> I wonder whether a VM with ReactOS would work for generating a Windows installer.

I'm not sure how we would use a VM on a service in the cloud that would build the PythonTurtle package in an automated fashion. Can you elaborate? (Do you intend for local use on a developer machine only?)

**Comment by cool-RR** [comment]
You're right, I didn't think this through. 

I guess there are 2 questions:

1. Could ReactOS work for generating a Windows installer using PyInstaller?

2. Could we set it up to work automatically, on the cloud?

Let's say that the answer to the first question is yes, and there isn't a SaaS provider that gives that service. Then say we can set up a Docker image that runs VirtualBox with a ReactOS guest. Maybe we can find a SaaS provider that'll run that Docker image for us?

**Comment by bittner** [comment]
https://github.com/pypa/python-packaging-user-guide/issues/267 contains quite a few interesting links to packaging Python applications.

**Comment by bittner** [comment]
As soon as uploading the Linux bundle to GitHub is settled (via PR #123) we'll jump on the [CI train for macOS](https://docs.travis-ci.com/user/reference/osx), yeah! :yum:

**Comment by cool-RR** [comment]
Sounds good to me!

**Comment by bittner** [comment]
As of now, building an application bundle on macOS (using PyInstaller) is not fully working. As an alternative, I thought, we may create a Homebrew formula. I'm currently [looking at the related documentation](https://docs.brew.sh/Python-for-Formula-Authors).

**Comment by bittner** [comment]
Just for reference, if the automatic releases don't work (they probably won't currently since the `api_key:` is set to [just a sample value](https://github.com/cool-RR/PythonTurtle/blob/master/.travis.yml#L74-L98)), here's how the [VSCodium project](https://github.com/VSCodium/vscodium/blob/master/.travis.yml#L31-L48) does automatic releases on GitHub. :+1:

**Comment by bittner** [comment]
Ram, you have mentioned "Windows installer" to be the top-priority of the present. I do agree.

Let's note down the current state of the union.

## Windows Installer (progress overview)

* Windows builds are [working on AppVeyor](https://ci.appveyor.com/project/cool-RR/PythonTurtle/history) since October 2018.
* Those builds create a Windows .EXE binary and a .MSI installer (see [example job](https://ci.appveyor.com/project/cool-RR/PythonTurtle)).
* The .MSI installer's file size is so small that it suggests that it's actually empty.
* I recall, you have tried out the .EXE (by downloading and running it) and you confirmed that .MSI wouldn't work. (Or was this someone else who tested it form me? Not sure.)
* I recall, I mentioned in some other issue (or was it a private chat of us?) what's needed to build a working MSI installer (it's actually just a MS Visual Studio project file, or so).

I think, I have prepared everything. Now, as I have neither a Windows computer nor I plan to buy one (same for macOS, by the way), the only thing missing is someone who takes over and puts the missing puzzle piece in place. From my gut feeling we're really, really close at "task completed".

## Next steps (volunteer needed)

It looks like, all you have to do is create the necessary configuration with the [Microsoft Visual Studio Installer Projects](https://marketplace.visualstudio.com/items?itemName=visualstudioclient.MicrosoftVisualStudio2017InstallerProjects) extension, according to [an answer on StackOverflow](https://stackoverflow.com/questions/44286704/create-setup-msi-installer-in-visual-studio-2017).

Hence, can you take over? What needs to be done, IIUC, is:

1. Create a Visual Studio Project with an Installer setup
1. Make sure the installer builds successfully on your local machine
1. Add the `vdproj` file to this project

AppVeyor will then detect that file, build the installer and provide it for download at the build [Artifacts](https://ci.appveyor.com/project/cool-RR/PythonTurtle/build/artifacts) page.

## References

* [Software pre-installed on Windows build VMs](https://www.appveyor.com/docs/windows-images-software/) (AppVeyor)
* [Added a Visual Studio Install Project and now and getting errors during build](https://help.appveyor.com/discussions/problems/626-added-a-visual-studio-install-project-and-now-and-getting-errors-during-build) (AppVeyor forum)
* [Example of building Microsoft Visual Studio 2017 Installer Projects](https://github.com/devlead/appveyor-vdproj-2017) (GitHub repo)

**Comment by cool-RR** [comment]
> From my gut feeling we're really, really close to "task completed".

I don't feel that if I'll start working on it, it'll be close to completed. I don't even know how to use Visual Studio Project, or what's our current plan. If I were to take this task, I'd guess it'd take a few days of full-time work.

You've put a lot of effort into this so I don't feel great about telling you this, but I definitely won't volunteer to work on this issue. PythonTurtle isn't close to being a priority in my life right now that I'd spend more than an hour, tops, on improving it. 

Am I being annoying by pointing to this issue from other issues, saying nothing is worth anything without this issue? I won't stop you from working on any other issues, I just probably wouldn't join because that effort seems wasted to me, but I won't block you, and if you'll need me for code reviews or small things, I'll be happy to do that.

**Comment by cool-RR** [comment]
Having another thought at this. You say "I have neither a Windows computer nor I plan to buy one". If your objection to do the items you listed is because you've done enough on this ticket, or that you prefer not to spend your time doing Windows-related things, then yes, that's a fair objection.

If the issue is that you don't want to spend money on a Windows license, then maybe that's solvable. One idea would be to use ReactOS, and seeing whether that runs Visual Studio Project. If ReactOS isn't an option, then maybe I could get a Windows license/machine to you. Let me know if that's helpful.

**Comment by bittner** [comment]
> Let me know if that's helpful.

Find a person that is _enthusiastic_ about programming on Windows. Enthusiastic about getting PythonTurtle running on a non-free, technically inferior operating system. That would be helpful.

That would add another person to be enthusiastic about bringing life to PythonTurtle, making it shine, pushing the project forward. If you don't want to spend time _programming_ maybe you want to spend your managerial skills to hire people, motivate contributors. That would help.

**Comment by bittner** [comment]
See https://github.com/PythonTurtle/PythonTurtle/issues/108#issuecomment-683420022 for alternative solutions to build an MSI installer (WiX, [WiX.Py](https://pypi.org/project/WiX.Py/)). We may be successful that way in combination with GitHub Actions. Does someone want to give that a shot?

**Comment by cool-RR** [comment]
> Does someone want to give that a shot?

I don't, so I guess we'll wait until a possible volunteer arrives.

**Comment by cclauss** [comment]
- [ ] __Linux__: pyinstaller build successful
    - [ ] Automated on GitHub Actions
- [ ] __macOS__: pyinstaller build successful
    - [ ] Automated on GitHub Actions
- [ ] __Windows__: pyinstaller build successful
    - [ ] Automated on GitHub Actions

**Comment by bittner** [comment]
A few hints to help you make progress:

- [x] With `python3 setup.py clean bundle` you can **build all binary bundles** (PyInstaller is used).
- [x] The binaries **build successful** for all platform (including .exe for Windows), for what I recall.
- [ ] For Windows users seem to expect a **Windows installer**; this could be a MSI or WiX.Py installer (both aren't yet prepared).
- [ ] GitHub Actions should **automatically push binaries** to our GitHub [Releases](https://github.com/PythonTurtle/PythonTurtle/releases) whenever we push a Git tag to `master`; we may want to use something like [actions/create-release](https://github.com/actions/create-release) to make this less laborious for us.

**Comment by bittner** [comment]
We received a super-nice answer by @judej on the Microsoft Developer Community [request for help](https://developercommunity.visualstudio.com/content/idea/940699/build-msi-installer-for-python-project-pythonturtl.html) that I had posted! :joy: 

Bottom line: No MSI volunteer for us, but there are two more interesting projects, one that builds an installer and the other that builds a [Python ZIP App](https://www.python.org/dev/peps/pep-0441/):

1. [pynsist](https://pypi.org/project/pynsist/) – _a tool to build Windows installers for your Python applications. \[...\] bundle Python itself, so you can distribute your application to people who don’t have Python installed_
1. [shiv](https://pypi.org/project/shiv/) – _for building fully self-contained Python zipapps as outlined in PEP 441, but with all their dependencies included!_

That could be two nice alternatives worth evaluating. Thanks Jude! :100: :medal_military:

**Comment by cool-RR** [comment]
Cool. I also recently heard about [PyOxidizer](https://github.com/indygreg/PyOxidizer). I'm still not 100% sure I understood the project ([discussion](https://github.com/indygreg/PyOxidizer/issues/289)) but it's possible it does what we want.

**Comment by cclauss** [comment]
I will take a shot at doing this with `shiv` over the weekend.  Please try `PyOxidizer` and then we can compare notes.

**Comment by cool-RR** [comment]
Sorry, I started a new job a few days ago and I won't have time to do anything :(

**Comment by bittner** [comment]
`PyOxidizer` sounds like an experiment to me: A Python application running on "a Rust VM", an OS-native layer created using Rust. I'm not sure what benefit this provides. Doesn't the `PyInstaller` setup we run do it all fine? :3rd_place_medal: 

`shiv` is a totally new approach (new in our "portfolio" of release artifacts), which might be interesting to provide to our audience. I find the idea thrilling (even though I'd be ready to bet few people might use it). :2nd_place_medal: 

`pynsist` sounds like the Microsoft Windows installer we're looking for. This might provide the user experience Windows folks seem to expect: :one: Download installer, :two: run it to install and add a start menu entry, :three: start application from start menu. :1st_place_medal:

**Comment by cclauss** [comment]
Downside of pynsist is windows only.  I will try to advance Shiv (from LinkedIn) on Mac and Linux.
